### PR TITLE
Make Prometheus namespace configurable

### PR DESCRIPTION
**Pull request by xulingqing**
The previous code hard coded Prometheus in istio-system namespace. 
This PR introduces a discoverPromNamespace function using labelSelector "app=prometheus"

### CODE REVIEW

File: watcher.go [hunk 1]
```diff
@@ -148,25 +149,52 @@ func (w *Watcher) waitForShutdown(stop <-chan struct{}) {
...
+
+func discoverPromNamespace(client *kubernetes.Clientset) (string, error) {
+	label := "prometheus"
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{"app": label}}
+	listOptions := metav1.ListOptions{
+		LabelSelector: labels.Set(labelSelector.MatchLabels).String(),
 	}
-	return configMap, nil
+
+	podsList, err := client.CoreV1().Pods("").List(context.TODO(), listOptions)
```
Comment (howardjohn) [inline watcher.go:191]: this operation is both expensive (listing all pods in cluster) and not accurate- there could be many Prometheus running.

Should we just have a config option?

Comment (xulingqing) [reply watcher.go:191]: I've tried list all namespace, but the client do not have permission to list all namespaces.  We only need to mount the fild-sd-config in one namespace and the watcher will keep watching all newly added vm.

The limitation of this implementation is the app name must be "prometheus".

For the config option, do you mean it's possible to include one more field to indicate prometheus namespace in the container config?

**Comment by xulingqing** [comment]
Current container log is like:
```
  $ k logs $(kubectl -n demo-attach-vm-ns1 get pods -l "app=prometheus" -o name) -n demo-attach-vm-ns1 -c vm-discovery
2021/06/02 20:40:39 starting new watcher for workload entries
2021/06/02 20:40:39 workload entry watcher created
2021/06/02 20:40:39 discover prometheus deployment in namespace demo-attach-vm-ns1
2021/06/02 20:40:39 waiting to be stopped
2021/06/02 20:40:39 VM workload demovm3-21.0.126.154-vmnetwork exists
2021/06/02 20:40:39 VM workload vm1-21.0.125.164-vmnetwork exists
2021/06/02 20:49:31 VM workload demovm3-21.0.126.154-vmnetwork exists
2021/06/02 20:49:31 VM workload demovm3-21.0.126.154-vmnetwork exists
2021/06/02 20:58:26 VM workload vm1-21.0.125.164-vmnetwork exists
2021/06/02 20:58:26 VM workload vm1-21.0.125.164-vmnetwork exists
2021/06/02 21:16:05 VM workload vm5-21.0.125.72-vmnetwork exists
2021/06/02 21:16:51 handle deleted workload 21.0.125.72
2021/06/02 21:16:51 Deleted VM workload vm5-21.0.125.72-vmnetwork

 $ k -n demo-attach-vm-ns1 exec $(kubectl -n demo-attach-vm-ns1 get pods -l "app=prometheus" -o name) -i -t -- cat etc/file_sd/staticConfigurations.json
Defaulting container name to prometheus-server-configmap-reload.
Use 'kubectl describe pod/prometheus-64c54c4c95-97qz2 -n demo-attach-vm-ns1' to see all of the containers in this pod.
[{"targets":["21.0.127.189:15020"]},{"targets":["21.0.126.154:15020"]},{"targets":["21.0.125.164:15020"]}]
```

File: watcher.go (hunk 1 above)
Comment (howardjohn) [reply watcher.go:191]: Yes. A user may have many prometheus instances, so automatically picking one seems wrong

Comment (douglas-reid) [reply watcher.go:191]: @howardjohn i think taking a config option is the right ultimate approach, but for default behavior, trying to find running prometheuses isn't all that problematic, as a one-time operation. 

@xulingqing can we move the ns discovery to a one-time startup operation (outside of the watch), and support multiple namespaces ?

File: Makefile [hunk 2]
```diff
@@ -1,7 +1,7 @@
 .PHONY: format docker
 
 BINARY=vm-discovery
-DOCKER_REPO=jackyzz
+DOCKER_REPO=linggg
```
Comment (douglas-reid) [inline Makefile]: @brian-avery can we find a place for these images that isn't an individual contributor repo?

File: watcher.go (hunk 1 above)
Comment (xulingqing) [reply watcher.go:191]: > @xulingqing can we move the ns discovery to a one-time startup operation (outside of the watch), and support multiple namespaces ?

do you mean create config map in all namespaces that have prometheus deployment?

File: Makefile (hunk 2 above)
Comment (brian-avery) [reply Makefile]: I created a docker organization -- `istioecosystem`. Lets use that instead.

Comment (xulingqing) [reply Makefile]: I tried to push to istioecosystem but got access denied error. Could you add me to the organization @brian-avery ? My docker id is `linggg`.

```
docker push istioecosystem/vm-discovery:latest
The push refers to repository [docker.io/istioecosystem/vm-discovery]
b01ed0cf0266: Preparing 
5f08512fd434: Preparing 
c7bb31fc0e08: Preparing 
50858308da3d: Preparing 
denied: requested access to the resource is denied
```

Comment (brian-avery) [reply Makefile]: Let me know when and I can build and upload it

File: watcher.go (hunk 1 above)
Comment (xulingqing) [reply watcher.go:191]: Done move the ns discovery to a one-time startup operation(at the beginning of `start` watcher function).

**Review by douglas-reid — APPROVED** [review]
Looks reasonable to me. I think we'll need to think about Operator and multiple instance scenarios in the future, but this seems OK for now, while we are stabilizing this bit of functionality.

Thanks for working on this!

File: Makefile (hunk 2 above)
Comment (xulingqing) [reply Makefile]: It's ready to build from master. Thanks @brian-avery

Comment (brian-avery) [reply Makefile]: Thanks and you're welcome :) https://hub.docker.com/r/istioecosystem/vm-discovery/tags

**Pull request merged**